modelforge/potential/neighbors.py

```diff
@@ -0,0 +1,418 @@
+
+            r_ij = (
+                torch.remainder(r_ij + box_lengths / 2, box_lengths) - box_lengths / 2
+            )
+
+        d_ij = torch.norm(r_ij, dim=1, keepdim=True, p=2)
+        return r_ij, d_ij
+
+
+class NeighborlistBruteNsq(torch.nn.Module):
+    """
+    Brute force N^2 neighbor list calculation for inference implemented fully in PyTorch.
+
+    This is compatible with TorchScript.
+
+
+    """
+
+    def __init__(
+        self,
+        cutoff: float,
+        displacement_function: OrthogonalDisplacementFunction,
+        only_unique_pairs: bool = False,
+    ):
+        """
+        Compute neighbor lists for inference, filtering pairs based on a cutoff distance.
+
+        Parameters
+        ----------
+        cutoff : float
+            The cutoff distance for neighbor list calculations.
+        displacement_function : OrthogonalDisplacementFunction
+            The function to calculate displacement vectors and distances between atom pairs, taking into account
+            the specified boundary conditions.
+        only_unique_pairs : bool, optional
+            Whether to only use unique pairs in the pair list calculation, by default False.
+        """
+
+        super().__init__()
+
+        self.register_buffer("cutoff", torch.tensor(cutoff))
+        self.register_buffer("only_unique_pairs", torch.tensor(only_unique_pairs))
+        self.displacement_function = displacement_function
+
+        self.indices = torch.tensor([])
+        self.i_final_pairs = torch.tensor([])
+        self.j_final_pairs = torch.tensor([])
+        log.info("Initializing Brute Force N^2 Neighborlist")
+
+    def forward(self, data: NNPInputTuple):
+        """
+        Prepares the input tensors for passing to the model.
+
+        This method handles general input manipulation, such as calculating
+        distances and generating the pair list. It also calls the model-specific
+        input preparation.
+
+        Parameters
+        ----------
+        data : NNPInputTuple
+            The input data provided by the dataset, containing atomic numbers,
+            positions, and other necessary information.
+
+        Returns
+        -------
+        PairListOutputs
+            Contains pair indices, distances (d_ij), and displacement vectors (r_ij) for atom pairs within the cutoff.
+        """
+        # ---------------------------
+        # general input manipulation
+        positions = data.positions
+        atomic_subsystem_indices = data.atomic_subsystem_indices
+
+        n = atomic_subsystem_indices.size(0)
+
+        # avoid reinitializing indices if they are already set and haven't changed
+        if self.indices.shape[0] != n:
+            # Generate a range of indices from 0 to n-1
+            self.indices = torch.arange(n, device=atomic_subsystem_indices.device)
+
+            # Create a meshgrid of indices
+            self.i_final_pairs, self.j_final_pairs = torch.meshgrid(
```

Comment:
does it make sense to downcast this to int32 and upcast when necessary (e.g. when using scatter operations)?

Reply:
I don't think that is necessary. this is not the bit that causes us to run out of memory as system size grows, and ultimately this will not be the routine we are going to use primarily (after we update NNPOps).

Reply:
I'd still prefer to use float32 and int32 wherever we can and upcast when necessary. Do you think that we might run in an overflow here that requires int64?

Reply:
I think we will run out of memory before we run out of indices for this algorithm. However, changing these to int32 right now will cause a cascade of a downstream failures at this point.  This might be good to explore outside this PR, to see if it has any really impact on efficiency and memory utilization; we would definitely need to put in a bunch of upcasting and probably need to define some global variable to set the dtype for indexing.

Reply:
Thanks for looking into this! I agree though, let's keep it at int64